scripts/west_commands/runners/intel_adsp.py

```diff
@@ -0,0 +1,130 @@
+        parser.add_argument('--rimage-tool', default=DEFAULT_RIMAGE_TOOL,
+                            help='path to the rimage tool')
+        parser.add_argument('--config-dir', default=DEFAULT_CONFIG_DIR,
+                            help='path to the toml config file')
+        parser.add_argument('--default-key',
+                            help='the default basename of the key store in board.cmake')
+        parser.add_argument('--key',
+                            help='specify where the signing key is')
+        parser.add_argument('--run-twister', action="store_true",
+                            help='specify the test is running by twister')
+
+    @classmethod
+    def do_create(cls, cfg, args):
+        return IntelAdspBinaryRunner(cfg,
+                                    remote_host=args.remote_host,
+                                    rimage_tool=args.rimage_tool,
+                                    config_dir=args.config_dir,
+                                    default_key=args.default_key,
+                                    key=args.key,
+                                    run_twister=args.run_twister
+                                    )
+
+    def do_run(self, command, **kwargs):
+        self.logger.info('Starting Intel ADSP runner')
+
+        # If the zephyr.ri is not existing, it will build and sign it.
+        if self.bin_fw is None or not os.path.isfile(self.bin_fw):
+            self.sign(**kwargs)
+
+        if re.search("intel_adsp_cavs", self.platform):
+            self.require(self.cavstool)
+            self.flash(**kwargs)
+        else:
+            self.logger.error("No suitable platform for running")
+            sys.exit(1)
+
+    def sign(self, **kwargs):
+        sign_cmd = ['west', 'sign', '-d', f'{self.cfg.build_dir}', '-t', 'rimage', '-p', f'{self.rimage_tool}',
+                '-D', f'{self.config_dir}', '--', '-k', f'{self.key}']
+
+        self.check_call(sign_cmd)
+
+    def flash(self, **kwargs):
+        # Generate a hash string for appending to the sending ri file
+        hash_object = hashlib.md5(self.bin_fw.encode())
+        random_str = f"{random.getrandbits(64)}".encode()
+        hash_object.update(random_str)
+        send_bin_fw = str(self.bin_fw + "." + hash_object.hexdigest())
+        os.rename(self.bin_fw, send_bin_fw)
+
+        # Copy the zephyr to target remote ADSP host and run
+        self.run_cmd = ([f'{self.cavstool}','-s', f'{self.remote_host}', f'{send_bin_fw}'])
+        self.log_cmd = ([f'{self.cavstool}','-s', f'{self.remote_host}', '-l'])
+        self.kill_cmd = ([f'{self.cavstool}','-s', f'{self.remote_host}', '-k'])
+
+        self.logger.debug(f"cavstool({self.cavstool}), fw('{send_bin_fw})")
+        self.logger.debug(f"rcmd: {self.run_cmd}")
+
+        self.check_call(self.run_cmd)
+
+        # When run by twister, it will not output the log
+        # message directly because the server's output log
+        # will grab by twister's device-serial-pty.
```

Comment:
This looks like a hack, we should not have special handling for twister, please consider a more generic solution or name this option differently.

Reply:
Yes, but I have some dilemma here. I would like to make "west flash"  here not only flash but also show the output message in directly STDOUT. This way is convenient I think,  though this behavior is not as the same as the devices which use the physical serial port. (west flash just flash without output log).  It needs a way to tell it runs in twister or not. Your suggestion is good, I propose to rename it to "--pty" because this is related the where the log output.

And I made another change there, the --pty generate by twister then pass to west flash. By this we just reduce the parameter to one, ex:   --west-flash="--remote-host=192.168.1.x".  How do you think?

Reply:
I found a problem here but might be solved a little later. So I still keep it --west-flash="--remote-host=192.168.1.x,--pty" here in twister. Thanks!